src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java

```diff
@@ -290,7 +248,11 @@ public ResponseEntity<byte[]> getFileUploadSubmission(@PathVariable Long exercis
             throw new AccessForbiddenException();
         }
 
-        return buildFileResponse(filePathService.actualPathForPublicPath(URI.create(submission.getFilePath())), false);
+        if (submission.getFilePath() == null) {
+            throw new EntityNotFoundException("Submission " + submissionId + " has no file");
+        }
+
+        return buildFileResponse(Objects.requireNonNull(FilePathService.actualPathForPublicPath(URI.create(submission.getFilePath()))), false);
```

Comment:
You repeat this check and path construction quite a few times. Is it possible to move (some of) this into buildFileResponse or another helper method?